src/python/gudhi/clustering/_tomato.cc

```diff
@@ -0,0 +1,277 @@
+/*    This file is part of the Gudhi Library - https://gudhi.inria.fr/ - which is released under MIT.
+ *    See file LICENSE or go to https://gudhi.inria.fr/licensing/ for full license details.
+ *    Author(s):       Marc Glisse
+ *
+ *    Copyright (C) 2020 Inria
+ *
+ *    Modification(s):
+ *      - YYYY/MM Author: Description of the modification
+ */
+
+#include <boost/container/flat_map.hpp>
+#include <boost/pending/disjoint_sets.hpp>
+#include <boost/property_map/property_map.hpp>
+#include <boost/property_map/transform_value_property_map.hpp>
+#include <boost/property_map/vector_property_map.hpp>
+#include <boost/property_map/function_property_map.hpp>
+#include <boost/iterator/counting_iterator.hpp>
+#include <boost/range/irange.hpp>
+#include <boost/range/adaptor/transformed.hpp>
+#include <vector>
+#include <unordered_map>
+#include <pybind11/pybind11.h>
+#include <pybind11/numpy.h>
+#include <iostream>
+
+namespace py = pybind11;
+
+template <class T, class = std::enable_if_t<std::is_integral<T>::value>>
+int getint(int i) {
+  return i;
+}
+// Gcc-8 has a bug that breaks this version, fixed in gcc-9
+// template<class T, class=decltype(std::declval<T>().template cast<int>())>
+// int getint(T i){return i.template cast<int>();}
+template <class T>
+auto getint(T i) -> decltype(i.template cast<int>()) {
+  return i.template cast<int>();
+}
+
+// Raw clusters are clusters obtained through single-linkage, no merging.
+
+typedef int Point_index;
```

Comment:
Indeed. I didn't try to benchmark on huge input. Looking for the modes of a function on a grid should be usable with a few billion points without taking too long. I'll think about using a 64-bit type...

Reply:
Looking for the modes of a 1D function:
```python
import numpy as np
from gudhi.clustering.tomato import Tomato

n=100000000
a=np.arange(0, 3*np.pi, 3*np.pi/n)
b=np.sin(a)+a/10
b=b+np.random.rand(n)/100
class N:
    def __getitem__(self,i):
        if i==0:
            return (1,)
        if i==n-1:
            return (i-1,)
        return (i-1,i+1)

t=Tomato('manual','manual')
t.fit(N(),weights=b)
```
This takes about 1 min for 10^8 points and grows rougly linearly with the number of points, so for this simple application it would take less than 30 minutes to reach the limit of 2 billions.